### PR TITLE
Patient Details Endpoints

### LINKED ISSUE #159: API endpoints to add patient details directly to MPI [MPIAPI]
## Summary
Add two new API endpoints that allow users to directly add/update patient record details in the MPI.

## Details / Tasks
`POST /patient -d '{"record": "...", "person": "person_reference_id"}'`
Add the above endpoint to allow users to directly add a patient record to the MPI, skipping the record linkage algorithm.

`PATCH /patient/<patient-reference-id> -d '{"record": "...", "person": "person_reference_id"}'`
Add the above endpoint to allow users to update an existing patient record in the MPI.
NOTE: For this operation, the `record` and `person` attributes are optional.  If not specified, the existing value will stay unchanged.  However, at least 1 must be specified.

## Related Issues/PRs
#155

## Notes / Comments
This completes use cases 9 and 10 in #155

### PR DESCRIPTION
**Pull request by cbrinson-rise8**
## Description
Add two new API endpoints that allow users to directly add/update patient record details in the MPI.

## Related Issues
closes #159 

## Additional Notes
`POST /patient -d '{"record": "...", "person": "person_reference_id"}'`
Add the above endpoint to allow users to directly add a patient record to the MPI, skipping the record linkage algorithm.

`PATCH /patient/<patient-reference-id> -d '{"record": "...", "person": "person_reference_id"}'`
Add the above endpoint to allow users to update an existing patient record in the MPI.
NOTE: For this operation, the `record` and `person` attributes are optional.  If not specified, the existing value will stay unchanged.  However, at least 1 must be specified.

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: src/recordlinker/routes/patient_router.py [hunk 1]
```diff
@@ -65,6 +66,73 @@ def update_person(
...
+    payload: typing.Annotated[schemas.PatientCreatePayload, fastapi.Body],
+    session: orm.Session = fastapi.Depends(get_session),
+) -> schemas.PatientRef:
+    """
+    Create a new patient record in the MPI
+    """
+    person = service.get_person_by_reference_id(session, payload.person_reference_id)
+
+    if person is None:
+        raise fastapi.HTTPException(status_code=fastapi.status.HTTP_422_UNPROCESSABLE_ENTITY)
```
Comment (m-goggins) [inline src/recordlinker/routes/patient_router.py:85]:
```suggestion
        raise fastapi.HTTPException(status_code=fastapi.status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Person not found. Must include a person_reference_id that already exists in the MPI.")
```

@ericbuckley How do you feel about more detailed error messages? I initially got a 422 when testing out locally because I hadn't merged in the changes from last week (the EO ones) and was confused about which part of the request was incorrect.

File: tests/unit/routes/test_patient_router.py [hunk 2]
```diff
@@ -66,6 +66,82 @@ def test_update_person(self, client):
...
+
+class TestCreatePatient:
+    def test_missing_data(self, client):
+        response = client.post("/patient")
+        assert response.status_code == 422
+
+    def test_invalid_person(self, client):
+        data = {"person_reference_id": str(uuid.uuid4()), "record": {}}
+        response = client.post("/patient", json=data)
+        assert response.status_code == 422
```
Comment (m-goggins) [inline tests/unit/routes/test_patient_router.py:78]:
```suggestion
        assert response.status_code == 422
        assert response.json() == {'detail': 'Person not found. Must include a person_reference_id that already exists in the MPI.'}

```

If more detailed errors are allowed.

File: src/recordlinker/routes/patient_router.py [hunk 3]
```diff
@@ -65,6 +66,73 @@ def update_person(
...
+    "/",
+    summary="Create a patient record and link to an existing person",
+    status_code=fastapi.status.HTTP_201_CREATED,
+)
+def create_patient(
+    payload: typing.Annotated[schemas.PatientCreatePayload, fastapi.Body],
+    session: orm.Session = fastapi.Depends(get_session),
+) -> schemas.PatientRef:
+    """
+    Create a new patient record in the MPI
```
Comment (m-goggins) [inline src/recordlinker/routes/patient_router.py:80]:
```suggestion
    Create a new patient record in the MPI and link to an existing person.
```

I was initially confused by this description because it doesn't imply that you need a `person_reference_id`. I thought this was a function that would just insert a new patient and assign them a `person_reference_id` (which we obviously should do but that's what I thought would happen based on the description alone).

File: src/recordlinker/routes/patient_router.py [hunk 4]
```diff
@@ -65,6 +66,73 @@ def update_person(
...
+    patient_reference_id: uuid.UUID,
+    payload: typing.Annotated[schemas.PatientUpdatePayload, fastapi.Body],
+    session: orm.Session = fastapi.Depends(get_session),
+) -> schemas.PatientRef:
+    """
+    Update an existing patient record in the MPI
+    """
+    patient = service.get_patient_by_reference_id(session, patient_reference_id)
+    if patient is None:
+        raise fastapi.HTTPException(status_code=fastapi.status.HTTP_404_NOT_FOUND)
```
Comment (m-goggins) [inline src/recordlinker/routes/patient_router.py:114]: Ditto more detailed error message here

File: src/recordlinker/routes/patient_router.py [hunk 5]
```diff
@@ -65,6 +66,73 @@ def update_person(
...
+    summary="Update a patient record",
+    status_code=fastapi.status.HTTP_200_OK,
+)
+def update_patient(
+    patient_reference_id: uuid.UUID,
+    payload: typing.Annotated[schemas.PatientUpdatePayload, fastapi.Body],
+    session: orm.Session = fastapi.Depends(get_session),
+) -> schemas.PatientRef:
+    """
+    Update an existing patient record in the MPI
```
Comment (m-goggins) [inline src/recordlinker/routes/patient_router.py:110]: For my own clarity, this operation can update the record (the PIIRecord) and/or the `person_reference_id` the existing patient is linked to, right?

**Review by m-goggins — COMMENTED** [review]
@ericbuckley Looks pretty good to me, though I think some of the docstrings and error messages could potentially be improved to make the new functionality clearer to users. It seems like the error messages in the codebase are fairly minimal so I'm not sure if that was a decision the team previously made. I tend to err on the side of more descriptive errors.

File: src/recordlinker/routes/patient_router.py (hunk 5 above)
Comment (ericbuckley) [reply src/recordlinker/routes/patient_router.py:110]: Correct, this is a `PATCH` operation, therefore every data element is optional.

File: src/recordlinker/routes/patient_router.py (hunk 4 above)
Comment (m-goggins) [reply src/recordlinker/routes/patient_router.py:117]: That's a fair point; I'm okay to add detail for 422s but leave 404s as they are.

File: src/recordlinker/routes/patient_router.py (hunk 5 above)
Comment (m-goggins) [reply src/recordlinker/routes/patient_router.py:113]: Right. I think the docstring could clarify it with something like "Update an existing patient record in the MPI by updating the contents of the PII Record and/or by updating `person_reference_id` the patient record is linked to." ...or something like that

Comment (ericbuckley) [reply src/recordlinker/routes/patient_router.py:113]: I was thinking the [differences between PUT and PATCH](https://www.geeksforgeeks.org/difference-between-put-and-patch-request/#difference-between-put-and-patch-request) imply that it's an and/or operation, but it doesn't hurt to be explicit.  If you think that's lost and/or confusing, we can certainly expand on the docstring.

File: tests/unit/routes/test_patient_router.py (hunk 2 above)
Comment (ericbuckley) [reply tests/unit/routes/test_patient_router.py:78]: we sure can, I'm going to make some adjustments on the format as I'd like the detail response to match that of other 422 validation errors

```json
"detail": [
  {
    "loc": ["string"],
    "msg": "string",
    "type": "string"
  }
]
```

**Review by m-goggins — APPROVED** [review]
